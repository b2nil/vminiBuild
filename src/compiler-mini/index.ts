export { compileScript } from "./script"
export { compileTemplate, normalizeAssetsURLOptions, getPlatformDirective } from "./template"
export { parse, compileStyleAsync } from "@vue/compiler-sfc"