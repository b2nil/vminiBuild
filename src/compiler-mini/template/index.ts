import { stringifyNodes } from './stringify'
import { transformSyntax, transformAssetUrls, transformStyleWithCssVars, transformCssModuleClasses } from './transforms'
import { baseParse, CompilerError, transform } from "@vue/compiler-core"

import type { VueOptions } from 'types'
import type { ElementNode } from '@vue/compiler-core'
import type {
  SFCTemplateCompileOptions,
  SFCTemplateCompileResults
} from '@vue/compiler-sfc'

interface PreProcessor {
  render (
    source: string,
    options: any,
    cb: (err: Error | null, res: string) => void
  ): void
}

function preprocess (
  preprocessor: PreProcessor,
  {
    source,
    filename,
    preprocessOptions
  }: Pick<SFCTemplateCompileOptions,
    "source" | "filename" | "preprocessOptions">
): string {
  let res: string = ''
  let err: Error | null = null

  preprocessor.render(
    source,
    { filename, ...preprocessOptions },
    (_err, _res) => {
      if (_err) err = _err
      res = _res
    }
  )

  if (err) throw err
  return res
}

export { getPlatformDirective } from "./utils"
export { normalizeAssetsURLOptions } from "./transforms"

export async function compileTemplate (
  template: string,
  options: VueOptions["template"] & {
    filename: string,
    id: string,
    cssVars: string[],
    isComponent?: boolean,
    cssModules: Record<string, Record<string, string>>,
    platformDir: string
  }
): Promise<SFCTemplateCompileResults> {
  const errors: CompilerError[] = []
  const warnings: CompilerError[] = []
  const preprocessLang = options?.preprocessLang

  if (preprocessLang && preprocessLang === "pug") {
    try {
      const pug = options.preprocessCustomRequire!("pug")
      template = preprocess(pug, {
        source: template,
        filename: options.filename,
        preprocessOptions: options.preprocessOptions
      })
      // Fix #default="#default" and v-else="v-else"
      template = template.replace(/(\B#.*?|\bv-.*?)="\1"/g, "$1")
    } catch (e: any) {
      return {
        code: `export default function render() {}`,
        source: template,
        tips: [],
        errors: [e]
      }
    }
  }

  const assetOptions = options.transformAssetUrls
    ? typeof options.transformAssetUrls === "boolean"
      ? {}
      : options.transformAssetUrls
    : {}

  const longId = `data-v-${options.id}`

  const ast = baseParse(template, options?.compilerOptions || {})
  transform(ast, {
    ...(options?.compilerOptions || {}),
    scopeId: options.scoped ? longId : undefined,
    nodeTransforms: [
      ...(options?.compilerOptions?.nodeTransforms || []),
      transformCssModuleClasses(options.cssModules),
      transformSyntax(options.platformDir),
      transformAssetUrls(assetOptions),
      transformStyleWithCssVars(options.id, options.cssVars, options.isComponent),
    ]
  })

  const code = stringifyNodes(ast.children, ast as unknown as ElementNode, options.platformDir)

  return {
    code,
    ast,
    source: template,
    tips: [],
    errors
  }
}