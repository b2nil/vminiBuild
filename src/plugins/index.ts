export * from "./cache"
export * from "./vue"
export * from "./style"
export * from "./entry"
export { printStats } from "./emits"