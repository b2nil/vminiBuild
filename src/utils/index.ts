export * from "./cache"
export * from "./paths"
export * from "./config"
export * from "./utils"

export const enum DEFINE {
  PROPS = "defineProps",
  EXPOSE = "defineExpose",
  HOOK = "defineHookConfig",
  APP_CONFIG = "defineAppConfig",
  PAGE_CONFIG = "definePageConfig",
  BUILD_CONFIG = "defineBuildConfig",
  PROJECT_CONFIG = "defineProjectConfig",
  SITEMAP_CONFIG = "defineSitemapConfig",
}