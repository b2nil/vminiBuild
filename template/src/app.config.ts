export default defineAppConfig({
  pages: [
    "pages/home/index",
    "pages/mine/index",
  ],
  subPackages: [
    {
      root: "subPackageA",
      pages: ["another/index"]
    },
  ],
  window: {
    navigationBarBackgroundColor: "#1e1e1e",
    navigationBarTextStyle: "white",
    navigationBarTitleText: "vue-mini",
    navigationStyle: "default"
  },
  tabBar: {
    color: "#000",
    selectedColor: "#168",
    backgroundColor: "#fff",
    list: [
      {
        text: "Home",
        pagePath: "pages/home/index",
        iconPath: "assets/home.png",
        selectedIconPath: "assets/home-selected.png"
      },
      {
        text: "Mine",
        pagePath: "pages/mine/index",
        iconPath: "assets/mine.png",
        selectedIconPath: "assets/mine-selected.png"
      }
    ]
  }
})