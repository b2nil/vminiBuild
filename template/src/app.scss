@import "./styles/mixins.scss";
@import "./styles/variables.scss";

page {
  color: $font-color;
  font-size: $font-size;
  font-family: sans-serif;
  @include lightBackground;
}
