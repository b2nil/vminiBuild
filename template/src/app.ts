//@ts-ignore
import PromisePolyfill from 'promise-polyfill'
import 'regenerator-runtime'
import { createApp } from '@vue-mini/wechat'

Promise = PromisePolyfill

createApp(() => {
  console.log('App Launched!')
})
