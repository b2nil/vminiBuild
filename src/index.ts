import fs from "fs"
import path from 'path'
import esbuild from "esbuild"
import { cac } from "cac"
import { customRequire, DEFINE, extractConfigFromFile } from './utils'
import {
  printStats,
  vuePlugin,
  styleLoader,
  nativePlugin,
  vueminiPlugin,
} from "./plugins"

import type { BuildOptions } from "esbuild"
import type { UserConfig, CliOptions } from 'types'
import { normalizeAssetsURLOptions } from './compiler-mini'

const cli = cac("vmini")

async function getUserBuildConfig (opts: CliOptions): Promise<UserConfig> {
  let configFile = opts.c || opts.config
  let resolvedPath: string | undefined

  if (configFile) {
    const specifiedPath = path.resolve(process.cwd(), configFile)
    if (fs.existsSync(specifiedPath)) {
      resolvedPath = specifiedPath
    }
  } else {
    const tsconfigFile = path.resolve(process.cwd(), `mini.config.ts`)
    if (fs.existsSync(tsconfigFile)) {
      resolvedPath = tsconfigFile
    }

    if (!resolvedPath) {
      const jsconfigFile = path.resolve(process.cwd(), `mini.config.js`)
      if (fs.existsSync(jsconfigFile)) {
        resolvedPath = jsconfigFile
      }
    }
  }

  const buildConfig: UserConfig = {
    outDir: `dist`,
    emptyOutDir: false,
    platform: "weapp",
    designWidth: 750,
    logLevel: "debug"
  }

  if (!resolvedPath) {
    console.warn(`config file mini.config.ts/js not found`)
    console.warn(`use default config`)
    return buildConfig
  }

  const config = await extractConfigFromFile(resolvedPath, DEFINE.BUILD_CONFIG)

  if (!config) {
    console.warn(`Failed to read build config at: ${resolvedPath}`)
    console.warn("Please use `export default` or `exports.default` to export the build config")
    console.warn("Now use the build-in config instead.")
    return buildConfig
  }

  return config as UserConfig
}

async function initBuildOptions (userConfig: UserConfig): Promise<BuildOptions> {
  const { useCDN, ...userBuildOptions } = userConfig
  userBuildOptions.vue = {
    ...(userBuildOptions.vue || {}),
    useCDN,
    style: {
      ...(userBuildOptions.vue?.style || {}),
      postcssPlugins:
        userBuildOptions.vue?.style?.postcssPlugins
          ? userBuildOptions.vue?.style?.postcssPlugins
          : [
            customRequire(`postcss-pxtransform`)({
              platform: userBuildOptions.platform || `weapp`,
              designWidth: userBuildOptions.designWidth || 750
            })
          ],
      preprocessCustomRequire: customRequire
    },
    template: {
      ...(userBuildOptions.vue?.template || {}),
      preprocessCustomRequire: customRequire
    }
  }

  if (useCDN) {
    let { transformAssetUrls } = userBuildOptions.vue.template!

    if (!transformAssetUrls || transformAssetUrls === true) {
      transformAssetUrls = {
        base: `http://${useCDN.host}:${useCDN.port}/`
      }
    } else {
      transformAssetUrls = normalizeAssetsURLOptions(transformAssetUrls, {
        base: `http://${useCDN.host}:${useCDN.port}/`,
        includeAbsolute: false,
        tags: {}
      })
    }

    userBuildOptions.vue.template!["transformAssetUrls"] = transformAssetUrls
  }


  const buildOptions: BuildOptions = {
    entryPoints: ["src/app.ts", "src/app.config.ts"],
    outdir: userBuildOptions.outDir || `dist`,
    target: "esnext",
    format: "esm",
    bundle: true,
    color: true,
    metafile: true,
    incremental: true,
    splitting: true,
    define: {
      ...(userBuildOptions.define || {}),
      "process.env.__PLATFORM__": JSON.stringify(userBuildOptions.platform || "weapp")
    },
    watch: !userBuildOptions.watch ? false : {
      onRebuild (error, _res) {
        if (error) console.error('[x] rebuild failed:', error.message, "\n", error.stack)
        else console.log('[+] rebuild succeeded')
      }
    },
    logLevel: userBuildOptions.logLevel,
    chunkNames: 'common/[dir]/[name]-[hash]',
    assetNames: 'assets/[dir]/[name]-[hash]',
    plugins: [
      vueminiPlugin({
        watch: userBuildOptions.watch,
        minify: userBuildOptions.minify,
        aliases: userBuildOptions.aliases
      }),
      vuePlugin(userBuildOptions.vue),
      styleLoader(userBuildOptions.vue),
      nativePlugin(userBuildOptions.vue),
      ...(userBuildOptions.plugins || [])
    ]
  }

  return buildOptions
}

async function buildWithEsbuild (options: CliOptions) {
  const userConfig = await getUserBuildConfig(options)

  if (options.p || options.platform) {
    userConfig.platform = options.p || options.platform
  }

  process.env.__PLATFORM__ = userConfig.platform || 'weapp'

  userConfig.watch = options.w || options.watch
  userConfig.logLevel = options.l || options.logLevel
  userConfig.emptyOutDir = options.emptyOutDir || userConfig.emptyOutDir
  userConfig.minify = options.minify || userConfig.minify

  const buildOptions = await initBuildOptions(userConfig)

  // temprary hack to serve the whole src dir
  if (userConfig.useCDN) {
    esbuild.serve({
      ...userConfig.useCDN,
      servedir: `src`
    }, {
      entryPoints: [],
      outdir: `src/${userConfig.useCDN?.servedir || "images"}`,
      write: false
    }).then(_server => {
      // do nothing
    })
  }

  esbuild.build(buildOptions)
    .then(async (res) => {
      if (buildOptions.watch || userConfig.useCDN) {
        if (userConfig.useCDN) {
          console.log(`[+] serving static files at: http://${userConfig.useCDN.host}:${userConfig.useCDN.port}`)
        }
        console.log('[+] watching...')
      } else {
        await printStats(res.metafile!, esbuild)
      }
    })
    .catch((e: any) => {
      console.error(`[x] build error: \n${e.stack || e}`)
    })
}

export function printVersion () {
  const root = path.resolve(__dirname, "../")
  const ver = require(path.join(root, "package.json")).version
  console.log(`(^o^) vminiBuild v${ver} (^o^)`)
}

export function run () {
  cli
    .option("-p, --platform [platform]", "[string] weapp | alipay | swan | tt", { default: "weapp" })
    .option("-l, --loglevel <level>", `[string] debug | warn`)
    .option("-c, --config <configFilePath>", `[string] specify config file path, defualt: mini.config.ts/js`)

  cli
    .command("build")
    .option("-w, --watch", "[boolean] enable watch mode")
    .option("--minify", "[boolean] force minify the outputs")
    .option("--emptyOutDir", "[boolean] force empty OutDir")
    .action(buildWithEsbuild)

  cli.help()
  cli.parse()
}
