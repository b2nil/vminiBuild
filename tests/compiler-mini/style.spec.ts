import { compileStyleAsync } from '../../src/compiler-mini'

const defaultOption = {
  id: "test",
  filename: "test.css"
}

async function compile (
  css: string,
  options: any = defaultOption
) {
  const res = await compileStyleAsync({
    source: css,
    ...options
  })
  return res
}

const css = `
.btn {
  color: red;
}
`

describe("transform styles", () => {
  it("should transform css module", async () => {
    const res = await compile(css, { ...defaultOption, modules: true })
    expect(res.code).toMatchSnapshot()
    expect(res.modules).toMatchSnapshot()
  })
})

