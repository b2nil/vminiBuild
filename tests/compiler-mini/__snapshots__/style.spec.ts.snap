// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform styles should transform css module 1`] = `
"
._btn_xhp46_2 {
  color: red;
}
"
`;

exports[`transform styles should transform css module 2`] = `
Object {
  "btn": "_btn_xhp46_2",
}
`;
